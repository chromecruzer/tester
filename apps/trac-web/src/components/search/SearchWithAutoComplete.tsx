import _ from "lodash";
import React, {useMemo, useRef, useState} from "react";
import {MdOutlineClear} from "react-icons/md";
import AutoCompleteDebouncer from "./AutoCompleteDebouncer";
import {SearchSpecifications, useTracContext} from "../../TracContext";
import {useQuery} from "@tanstack/react-query";
import {postSearchForSuggestions} from "../../api/searchApi";
import {NullableString} from "@trac/datatypes";
const debouncer = AutoCompleteDebouncer.createInstance(3000);
export type ArrowKeys = ('ArrowDown' | 'ArrowUp' | 'PageDown' | 'PageUp' | null );
const pageSize = 5;

export const SearchWithAutoComplete = ({setChosen, label = 'Search', alwaysSuggestions = false,
                                         children}) => {
  const typedText = useRef<HTMLInputElement>(null);
  const ulRef = useRef<HTMLUListElement>(null);
  const {getTracContext} = useTracContext();
  const {filters: filtersFromContext} = getTracContext("search") as SearchSpecifications;
  const [searchText, setSearchText] = useState(null as NullableString);
  const [inputHasFocus, setInputHasFocus] = useState(true);
  const [suggestionsCursor, setSuggestionsCursor] = useState(0);
  const scrollOffsets = {
    ArrowUp: 1,
    ArrowDown: 0,
    PageUp: 2,
    PageDown: -1,
  }
  const cursorMath = (change: ArrowKeys) => {
    const increment = change === 'ArrowUp' || change === 'ArrowDown' ? 1 : pageSize;
    const direction = change === 'ArrowUp' || change === 'PageUp' ? -1 : 1;
    const newCursor = suggestionsCursor + (increment * direction);
    console.log(`new cursor ${newCursor} from suggestions cursor ${suggestionsCursor}, increment ${increment} and direction ${direction}`)
    if(newCursor < 0) {
      return 0;
    }
    if(newCursor >= suggestions.length) {
      return suggestions.length - 1;
    }
    return newCursor;
  }
  // console.log(`autocomplete search ${searchText}`, filters);
  const setSearch = text => {
    console.log(`AUTO complete setting search text to ${text}`);
    setSearchText(text)

  }
  const getTypedText = () => {
    if(typedText.current) {
      return typedText.current.value;
    }
    return null;
  };
  const clearSearch = () => {
    // console.log('clearing the search');
    setSearch(null);
    if(typedText.current) {
      typedText.current.value = '';
    }
    debouncer.clear();
  }
  const onKeyChange = e => {
    const key = e.key;
    const text = getTypedText();
    // console.log(`processing ${key} and ${text}`,e);
    switch (key) {
      case 'Backspace':
      case 'Delete':
        if(!_.isString(text) || text.length === 1) {
          clearSearch();
          return;
        }
        break;
      case 'Enter':
      case 'Tab':
        if(inputHasFocus && !alwaysSuggestions && !_.isEmpty(suggestions)) {
          console.log(`choosing typed text `, getTypedText());
          setChosen({field: getTypedText(), suggestion: null});
        } else {
          console.log(`choosing suggestion ${suggestionsCursor} at`, suggestions[suggestionsCursor]);
          setChosen({field: null, suggestion: suggestions[suggestionsCursor]})
        }
        break;
      case 'ArrowDown':
      case 'ArrowUp':
      case 'PageDown':
      case 'PageUp':
        // console.log(`setting suggestions key for cursor ${suggestionsCursor}`, key)
        setSuggestionsCursor(() => cursorMath(key));
        if(inputHasFocus) {
          setInputHasFocus(false);
        }
        console.log('UL Children', ulRef.current?.children)
        ulRef.current?.children[suggestionsCursor + scrollOffsets[key]].scrollIntoView();
        console.log('UL Children', ulRef.current?.children[suggestionsCursor + scrollOffsets[key]])
        break;
      default:
        debouncer.set(() => setSearch(`${text}${key}`), `${text}${key}`, `${text}${key}`.length);
    }
  }
  const {
    isLoading,
    isError,
    error,
    data: suggestions
  } = useQuery(['search', searchText, filtersFromContext],
    () => postSearchForSuggestions(searchText, filtersFromContext), {
      select: data => {
        // console.log(`querying using ${searchText}`, filtersFromContext);
        // console.log('queried suggestions', data)
        return data
      },
      enabled: !!searchText,
    })
  if (isError) {
    console.error(`Server returned an error ${error}`)
  }
  const onClick = e => {
    // console.log(`clicked on`, e);
    const selected = _.find(suggestions, s => s.field === e.target.firstChild.textContent)
    setChosen({field: null, suggestion: selected});
    clearSearch();
  }

  const suggestionHasFocus = (index) => {
    // console.log(`setting suggestions index to ${index}`);
    setInputHasFocus(false)
    setSuggestionsCursor(index)
  }

  const generateClassName = (index) => {
    // console.log(`setting highlight on index ${index}`)
    return index === suggestionsCursor ? 'bg-bl-consumer-second' : 'bg-bl-consumer-light';
  }
  // console.log(`visible with text ${searchText} and loading ${isLoading}`)
  const suggestionsList = useMemo(() => {
    if (getTypedText()) {
      switch (true) {
        case isLoading:
          return <ul>
            <li>...Loading...</li>
          </ul>;
        case _.isEmpty(suggestions):
          return <ul>
            <li>Nothing Was Found</li>
          </ul>;
        default:
          // console.log(`cursor at ${suggestionsCursor}`)
          return <ul className='overflow-y-auto h-32 border-2' ref={ulRef}>
            {_.map(suggestions, (s, i) => {
              // console.log(`generating selection for ${s.field} className ${generateClassName(i)}`);
              return <li
                key={s.uuid}
                onClick={onClick}
                onKeyDown={onKeyChange}
                data-tip={true}
                onMouseEnter={() => suggestionHasFocus(i)}
                className={generateClassName(i)}>
                {s.field}
              </li>
            })}
          </ul>
      }
    }
    return null;
  },[searchText, typedText.current, isLoading, suggestions, suggestionsCursor])


  return <div className='flex justify-s items-start'>
    <h4 className='py-2 px-3'>{label}</h4>
    <div className='flex-row'>
      <div className='flex border-2 hover: border-lime-500'>
        <input
          type='text'
          ref={typedText}
          onClick={() => setInputHasFocus(true)}
          placeholder='Search'
          // onChange={onChange}
          onKeyDown={onKeyChange}
          className='py-2 px-3 flex-grow cursor-pointer bg-bl-text-light text-bl-text-grey'
        />
        {getTypedText() !== null && <button onClick={clearSearch}><MdOutlineClear/>|</button>}
      </div>
      {suggestionsList}
    </div>
    {children}
  </div>
}
