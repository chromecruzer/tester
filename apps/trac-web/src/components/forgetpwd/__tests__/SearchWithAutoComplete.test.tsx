import {render, screen} from "@testing-library/react";
import React, {useState as useStateMock} from "react";
import * as reactRedux from "../../../redux/hooks";
import * as reactRouterDom from "react-router-dom";
import * as reactRouter from "react-router";
import userEvent from "@testing-library/user-event";
import {SearchWithAutoComplete} from "../SearchWithAutoComplete";
import util from "util";
import ManageSearchAutoCompleteState from "../ManageSearchAutoCompleteState";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

jest.mock('react-redux', () => ({
  useDispatch: jest.fn().mockReturnValue(jest.fn().mockReturnValue(jest.fn())),
  useSelector: jest.fn(),
}));
jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn(),
  useState: jest.fn(),
}))
jest.mock('../../../redux/searchActions', () => ({
  retrieveSuggestions: jest.fn(),
  searchClear: jest.fn(),
}));
jest.mock('../SearchSuggestions', () => ({
  SearchSuggestions: props => (<>
    <h4>Mock Suggestions</h4>
    <ul>
      <li key='selectSuggestion'>selectSuggestion is {props['selectSuggestion'] ? 'present' : 'absent'}</li>
      <li key='suggestions'>suggestions {props['suggestions'] ? ' are present' : ' are missing'}</li>
    </ul>
  </>),
}))
jest.mock('../SearchFilters', () => ({
  SearchFilters: props => (<>
    <h4>Mock Filter Buttons</h4>
    <ul>
      <li key='changeFilters'>changeFilters is {props['changeFilters'] ? 'present' : 'absent'}</li>
      <li key='buttons'>buttons {props['buttons'] ? ' is present' : ' is missing'}</li>
    </ul>
  </>),
}))

jest.mock('../SearchFilterButtonsState');
jest.mock('../ManageSearchAutoCompleteState');
// console.log(`mocked class ${dump(ManageSearchAutoCompleteState)}`);
// const {default: ManageSearchAutoCompleteStateMockConstructor, setTextParamMock,
//   setFilterParamsMock, setResultsMock, resetMock, changeSearchMock} =
//   (ManageSearchAutoCompleteState as unknown) as typeof ManageSearchAutoCompleteStateMock;
describe('SearchWithAutoComplete', () => {
  let setSearchResults, navigateSpy, useSearchParamsMock, mockReset, mockChangeSearch;;
  const renderSearchWithAutoComplete = () => {
    return render(<SearchWithAutoComplete
      setSearchResults={setSearchResults}
    />)
  }
  beforeEach(() => {
    navigateSpy = jest.fn();
    jest.spyOn(reactRouter, 'useNavigate').mockImplementation(() => navigateSpy);
    useSearchParamsMock = jest.spyOn(reactRouterDom, 'useSearchParams')
      .mockImplementation(() => ([new URLSearchParams(), jest.fn()]));
    jest.spyOn(reactRedux, 'useAppSelector').mockImplementation(() => ({
      loading: false,
      suggestions: [
        'suggestion 1', 'suggestion2'
      ],
      error: ''
    }));
    setSearchResults = jest.fn();
    mockReset =jest.spyOn(ManageSearchAutoCompleteState.prototype, 'reset');
    mockChangeSearch = jest.spyOn(ManageSearchAutoCompleteState.prototype, 'changeSearch');
  });
  it('should return a search box with filter buttons', () => {
    (useStateMock as jest.Mock)
      .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()]))
      .mockImplementationOnce(() => ([null, jest.fn()]))
      .mockImplementationOnce(() => ([false, jest.fn()]))
      .mockImplementationOnce(() => ([false, jest.fn()]));
    renderSearchWithAutoComplete();
    screen.getByText('Search');
    screen.getByRole('textbox');
    screen.getByText('Mock Filter Buttons');
    screen.getByText('changeFilters is present');
    screen.getByText('buttons is present');

  });
  it('should render a suggestions list when user has typed something', () => {
    (useStateMock as jest.Mock)
      .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()])) //filters
      .mockImplementationOnce(() => (['search text', jest.fn()])) // text
      .mockImplementationOnce(() => ([false, jest.fn()])) //results visible
      .mockImplementationOnce(() => ([true, jest.fn()])); // suggestions visible
    renderSearchWithAutoComplete();
    screen.getByText('Mock Suggestions');
    screen.getByText('selectSuggestion is present');
    screen.getByText('suggestions are present');
  });
  describe('onKeyChange', () => {
    it('should set the results when the key is a return or a tab', () => {
      const [searchTextFn, resultsFn, suggestionsFn] = [jest.fn(), jest.fn(), jest.fn()];
      userEvent.setup();
      (useStateMock as jest.Mock)
        .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()])) //filters
        .mockImplementationOnce(() => (['search text', searchTextFn])) // text
        .mockImplementationOnce(() => ([false, resultsFn])) //results visible
        .mockImplementationOnce(() => ([true, suggestionsFn])); // suggestions visible
      renderSearchWithAutoComplete();
      const box = screen.getByRole('textbox');
      return userEvent.type(box, '{Enter}').then(() => {
        console.log('userEvent finished')
        expect(navigateSpy.mock.calls).toMatchSnapshot();
      })
    });
    it('should clear the search results and set suggestions to true for any other key', () => {
      const [searchTextFn, resultsFn, suggestionsFn] = [jest.fn(), jest.fn(), jest.fn()];
      userEvent.setup();
      (useStateMock as jest.Mock)
        .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()])) //filters
        .mockImplementationOnce(() => (['pr', searchTextFn])) // text
        .mockImplementationOnce(() => ([false, resultsFn])) //results visible
        .mockImplementationOnce(() => ([false, suggestionsFn])); // suggestions visible
      renderSearchWithAutoComplete();
      const box = screen.getByRole('textbox');
      return userEvent.type(box, 'o').then(() => {
        expect(suggestionsFn).toBeCalledWith( true);
      })
    });
  });
  describe('onChange', () => {
    it('should clear the results of there is no search text', () => {
      const [searchTextFn, resultsFn, suggestionsFn] = [jest.fn(), jest.fn(), jest.fn()];
      userEvent.setup();
      (useStateMock as jest.Mock)
        .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()])) //filters
        .mockImplementationOnce(() => (['p', searchTextFn])) // text
        .mockImplementationOnce(() => ([false, resultsFn])) //results visible
        .mockImplementationOnce(() => ([false, suggestionsFn])); // suggestions visible
      renderSearchWithAutoComplete();
      const box = screen.getByRole('textbox');
      return userEvent.type(box, '{Backspace}').then(() => {
        expect(mockReset).toBeCalled();
      })
    });
    it('should set up to display a suggestions list when user has typed a first character', () => {
      userEvent.setup();
      (useStateMock as jest.Mock)
        .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()])) //filters
        .mockImplementationOnce(() => ([null, jest.fn()])) // text
        .mockImplementationOnce(() => ([false, jest.fn()])) //results visible
        .mockImplementationOnce(() => ([false, jest.fn()])); // suggestions visible
      renderSearchWithAutoComplete();
      const box = screen.getByRole('textbox');
      return userEvent.type(box, 'p').then(() => {
        expect(mockChangeSearch).toBeCalledWith('p', false);
      })
    });
    it('should set up to display a suggestions list when user has added to the text', () => {
      userEvent.setup();
      (useStateMock as jest.Mock)
        .mockImplementationOnce(() => ([{filter1: true, filter2: false}, jest.fn()])) //filters
        .mockImplementationOnce(() => (['pr', jest.fn()])) // text
        .mockImplementationOnce(() => ([false, jest.fn()])) //results visible
        .mockImplementationOnce(() => ([false, jest.fn()])); // suggestions visible
      renderSearchWithAutoComplete();
      const box = screen.getByRole('textbox');
      return userEvent.type(box, 'o').then(() => {
        expect(mockChangeSearch).toBeCalledWith('pro', false);
      })
    });
  });
});
