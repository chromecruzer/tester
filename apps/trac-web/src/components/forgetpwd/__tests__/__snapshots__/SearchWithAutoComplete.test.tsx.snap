// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchWithAutoComplete onKeyChange should set the results when the key is a return or a tab 1`] = `
Array [
  Array [
    "/?text=search text&filters=undefined",
  ],
]
`;
